### PR TITLE
[WIP] bringing code forward to django 2.0 compatibility

### PR DESCRIPTION
**Pull request by fgregg**
This PR brings this library to  django 2.0 compatibility.

### CODE REVIEW

**Comment by fgregg** [comment]
@schinckel thank you very much for writing this library.

I'n this branch, I'm working on getting the code to work with Django 2. It seems to be working correctly with Django 2.0 and Python 3.7, but the tests are failing in way that I don't understand.

I'm not sure if you are interested in updating your library to work with Django 2. If so, I wonder if you have some insight into these failing tests?

```bash
> tox
GLOB sdist-make: /Users/fgregg/experiment/django-proxy-overrides/setup.py
py37-django20 inst-nodeps: /Users/fgregg/experiment/django-proxy-overrides/.tox/.tmp/package/1/django-proxy-overrides-0.1.0.zip
py37-django20 installed: Django==2.0.9,django-proxy-overrides==0.1.0,pytz==2018.7
py37-django20 run-test-pre: PYTHONHASHSEED='3326298716'
py37-django20 runtests: commands[0] | coverage run -a --branch --source proxy_overrides runtests.py
WARNING: test command found but not installed in testenv
  cmd: /Users/fgregg/experiment/django-proxy-overrides/.env/bin/coverage
  env: /Users/fgregg/experiment/django-proxy-overrides/.tox/py37-django20
Maybe you forgot to specify a dependency? See also the whitelist_externals envconfig setting.
Creating test database for alias 'default'...
Traceback (most recent call last):
  File "runtests.py", line 44, in <module>
    runtests()
  File "runtests.py", line 40, in runtests
    execute_from_command_line(argv)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/commands/test.py", line 26, in run_from_argv
    super().run_from_argv(argv)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/base.py", line 316, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/base.py", line 353, in execute
    output = self.handle(*args, **options)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/commands/test.py", line 56, in handle
    failures = test_runner.run_tests(test_labels)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/test/runner.py", line 604, in run_tests
    old_config = self.setup_databases()
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/test/runner.py", line 551, in setup_databases
    self.parallel, **kwargs
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/test/utils.py", line 174, in setup_databases
    serialize=connection.settings_dict.get('TEST', {}).get('SERIALIZE', True),
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/db/backends/base/creation.py", line 68, in create_test_db
    run_syncdb=True,
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/__init__.py", line 148, in call_command
    return command.execute(*args, **defaults)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/base.py", line 353, in execute
    output = self.handle(*args, **options)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/core/management/commands/migrate.py", line 163, in handle
    pre_migrate_apps = pre_migrate_state.apps
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/utils/functional.py", line 37, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/db/migrations/state.py", line 210, in apps
    return StateApps(self.real_apps, self.models)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/db/migrations/state.py", line 271, in __init__
    self.render_multiple(list(models.values()) + self.real_models)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/db/migrations/state.py", line 306, in render_multiple
    model.render(self)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/db/migrations/state.py", line 574, in render
    return type(self.name, bases, body)
  File "/Users/fgregg/experiment/django-proxy-overrides/.env/lib/python3.7/site-packages/django/db/models/base.py", line 208, in __new__
    base.__name__,
django.core.exceptions.FieldError: Local field 'integer' in class 'BazProxy' clashes with field of the same name from base class 'Baz'.
ERROR: InvocationError for command '/Users/fgregg/experiment/django-proxy-overrides/.env/bin/coverage run -a --branch --source proxy_overrides runtests.py' (exited with code 1)
____________________________ summary _____________________________
ERROR:   py37-django20: commands failed
```

**Comment by schinckel** [comment]
Hi!

I think I stopped using it because in 1.11 it had the same problem: django became a lot stricter about overriding fields.

**Comment by fgregg** [comment]
Okay, thanks for the perspective, @schinckel. The funny thing is that it's working very well in a project for me, despite the failing tests.

File: proxy_overrides/base.py [hunk 1]
```diff
@@ -13,28 +13,32 @@ def override_model_field(model, name, field):
...
         if related_model._meta.proxy:
             raise TypeError('There is already a proxy model {!r} related to {!r} using {!r}'.format(
                 related_model.__name__,
-                field.rel.to.__name__,
+                field.remote_field.model.__name__,
                 related_name
             ))
 
     model.add_to_class(name, field)
+    model._meta.local_fields.remove(field)
```
Comment (schinckel) [inline proxy_overrides/base.py]: Oh, that is sneaky!

Comment (fgregg) [reply proxy_overrides/base.py]: yes! with this change, the tests run, makemigrations and systems-check does not complain.

the proxyforeignkey code works (which is what I really care about) but proxyfield does not work.

```bash
>tox
GLOB sdist-make: /Users/fgregg/experiment/django-proxy-overrides/setup.py
py37-django20 inst-nodeps: /Users/fgregg/experiment/django-proxy-overrides/.tox/.tmp/package/1/django-proxy-overrides-0.1.0.zip
py37-django20 installed: Django==2.0.9,django-proxy-overrides==0.1.0,pytz==2018.7
py37-django20 run-test-pre: PYTHONHASHSEED='2402096153'
py37-django20 runtests: commands[0] | coverage run -a --branch --source proxy_overrides runtests.py
WARNING: test command found but not installed in testenv
  cmd: /Users/fgregg/experiment/django-proxy-overrides/.env/bin/coverage
  env: /Users/fgregg/experiment/django-proxy-overrides/.tox/py37-django20
Maybe you forgot to specify a dependency? See also the whitelist_externals envconfig setting.
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
F.......
======================================================================
FAIL: test_override_compatible (tests.test_field.TestFieldOverride)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/fgregg/experiment/django-proxy-overrides/tests/test_field.py", line 15, in test_override_compatible
    self.assertEquals(3, baz_proxy.integer)
AssertionError: 3 != 2

----------------------------------------------------------------------
Ran 8 tests in 0.009s

FAILED (failures=1)
Destroying test database for alias 'default'...
ERROR: InvocationError for command '/Users/fgregg/experiment/django-proxy-overrides/.env/bin/coverage run -a --branch --source proxy_overrides runtests.py' (exited with code 1)
____________________________ summary _____________________________
ERROR:   py37-django20: commands failed
```

**Comment by fgregg** [comment]
Hi @schinckel,

For my purposes, all I need are the proxyforeignkey to work. I need this as a requirement for a library, so I need to it to be on pypi.

Would you like to adopt these changes, or should I fork and create a new library. Maybe django-proxy-foreignkey-overrides?

**Comment by schinckel** [comment]
Apparently I don't own the project that is actually on PyPI.

Maybe contact the owner and see if they will give you write access.

**Comment by schinckel** [comment]
https://pypi.org/project/django-proxy-overrides/

**Comment by schinckel** [comment]
@bashu You appear to own this: can you communicate with @fgregg your likelihood of giving him write access to the PyPI for this project?

**Comment by bashu** [comment]
@schinckel @fgregg done!

**Pull request merged**